feat: preferred_zones configuration (#1674)
aiven/aiven-docs: docs/products/kafka/howto/configure-preferred-zones.md

---
title: Configure preferred availability zones for Aiven for Apache Kafka®
sidebar_label: Configure preferred zones
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import ConsoleLabel from "@site/src/components/ConsoleIcons"
import ConsoleIcon from "@site/src/components/ConsoleIcons"
import RelatedPages from "@site/src/components/RelatedPages";

Configure preferred availability zones to control where your Aiven for Apache Kafka® service nodes are placed within a cloud region.

## Overview

> By default, Aiven automatically distributes Kafka service nodes across all available

Reviewer: ```suggestion
By default, Aiven automatically distributes Kafka service nodes across any available
```

Reviewer: To me, "all" implies we fill the region, e.g. 6 AZ example, we only deploy 3.

availability zones (AZs) in a region for maximum fault tolerance. With preferred
zones, you can restrict node placement to specific AZs while maintaining high
availability.

Use cases for preferred zones include:

- **Latency optimization**: Place nodes closer to your application workloads.
- **Cost reduction**: Minimize cross-AZ data transfer costs by co-locating Kafka

>   nodes with your consumers.

Reviewer: ```suggestion
  nodes with your applications.
```

Reviewer: Id probably also put this as the first bullet.

> - **Compliance requirements**: Restrict data to specific zones within a region.

Reviewer: ```suggestion
- **Compliance requirements**: Restrict data and deployments to specific zones within a region.
```

- **Follower fetching optimization**: Align broker placement with consumer locations
  for efficient rack-aware fetching.

## Prerequisites

> - An Aiven for Apache Kafka service running on AWS, Google Cloud, or Azure.

Reviewer: @harshini-rangaswamy do we need the R mark on every time we name or just for titles?

Reviewer 2: We only need the trademark symbol on the first occurrence of the product name.

> - An Aiven for Apache Kafka service running on AWS, Google Cloud, or Azure.

Reviewer: Is this just Kafka, or also applies to other service types?

Author: For now just kafka has the user_config, but we could also support other services in the future

- Knowledge of available zone IDs for your cloud region.
- [Aiven CLI](/docs/tools/cli) (optional).

## Zone ID formats

Zone IDs vary by cloud provider:

| Cloud provider | Format | Examples |
|---------------|--------|----------|
| AWS | Zone ID (account-independent) | `use1-az1`, `use1-az2`, `euc1-az1` |
| Google Cloud | Zone name | `europe-west1-a`, `us-central1-b` |
| Azure | Location/zone number | `germanywestcentral/1`, `westeurope/2` |

> :::note[AWS zone IDs]

Reviewer: Lets confirm if just AWS or also a Google issue, I think its also something on Google, Juha shared something the other day on it. We already reference on the doc site somewhere.

Author: Google AZ names should be stable across accounts

  --header 'Authorization: Bearer API_TOKEN' \
  --header 'content-type: application/json' \
  --data '{
    "user_config": {
      "preferred_zones": ["use1-az1", "use1-az2", "use1-az3"]
    }
  }'
```

Parameters:

- `PROJECT_NAME`: Name of your Aiven project.
- `SERVICE_NAME`: Name of your Aiven for Apache Kafka service.
- `API_TOKEN`: Your Aiven API token.

</TabItem>
<TabItem value="terraform" label="Terraform">

Use the `preferred_zones` attribute in your `aiven_kafka` resource:

```hcl
resource "aiven_kafka" "example" {
  project      = "my-project"
  service_name = "my-kafka"
  cloud_name   = "aws-us-east-1"
  plan         = "business-4"

  kafka_user_config {
    preferred_zones = ["use1-az1", "use1-az2", "use1-az3"]
  }
}
```

</TabItem>
</Tabs>

## How preferred zones work

When you configure preferred zones:

1. **New nodes** are placed in the specified zones when capacity is available.
1. **Zone validation** rejects invalid zone IDs at configuration time.
1. **Fallback behavior**: If a preferred zone is temporarily unavailable (for example,
   due to capacity constraints), nodes are placed in other available zones to maintain
   service availability.

> 1. **Existing nodes** are not automatically moved. Changes take effect on the next

Reviewer: ```suggestion
1. **Existing nodes** are not automatically rebalanced, unless they meet the conditions described below. Preferred zone creation is re-attempted on the next
```

   node recreation, such as during maintenance or plan changes.

### Minimum zone count

For high availability, configure at least three preferred zones. Configuring fewer

> than three zones requires special account permissions and reduces fault tolerance.

Reviewer: ```suggestion
than three zones will fail, and requires special account permissions to reduce fault tolerance.
```

Reviewer: Please confirm if it fails, if so add suggestion.

Reviewer: @keejon can you confirm.

Author: yes it fails if ACL is not present 👍

### Interaction with single zone configuration

> If both `preferred_zones` and `single_zone.availability_zone` are configured,

Reviewer: Could link out to the single zone setting, if documented already @harshini-rangaswamy ?

Reviewer 2: https://aiven.io/docs/products/kafka/reference/advanced-params-inkless#single_zone.availability_zone

the `single_zone` setting takes precedence when `single_zone.enabled` is true.

> ## Automatic rebalancing for Inkless Kafka

Reviewer: ```suggestion
## Automatic rebalancing for Standard plans
```

Reviewer: Assumed name change confirmed.

> For Inkless Kafka services with professional plans, nodes placed outside preferred

Reviewer: ```suggestion
For Kafka services with Standard professional plans, nodes placed outside preferred
```

> zones are automatically rebalanced once per day. This ensures that temporary zone

Reviewer: ```suggestion
zones are automatically re-attempted once per day. This attempts to fix any temporary issues of zone
```

unavailability during node creation is eventually corrected without manual
intervention.

Classic Kafka services receive alerts about misplaced nodes but do not automatically
rebalance. You can manually trigger node replacement through the Aiven Console or
API if needed.

## Example: Optimizing for follower fetching

To optimize costs with
[follower fetching](/docs/products/kafka/howto/enable-follower-fetching),
align your preferred zones with your consumer locations:

1. Identify the AZs where your Kafka consumers run.
1. Configure preferred zones to match those AZs.
1. Enable follower fetching on your Kafka service.
1. Configure `client.rack` on your consumers to match their AZ.

This setup ensures brokers and consumers are co-located, enabling consumers to fetch
from local replicas and reducing cross-AZ network costs.

> ## Example: Reducing cross-AZ costs for Inkless producers

Reviewer: ```suggestion
## Example: Reducing cross-AZ costs when producing to diskless topics
```

> For Inkless Kafka services with diskless topics, you can reduce cross-AZ data

Reviewer: ```suggestion
For Kafka services with diskless topics, you can reduce cross-AZ data
```

transfer costs by configuring producers to route requests to brokers in their local

> AZ. This uses the `client.id` pattern to communicate the producer's availability

Reviewer: Note we are developing enhanced functionality for this, using the ports.
Nothing for now but we can add to this later.

Diskless topics store data in object storage, which is accessible from all AZs. Any
broker can serve any partition. By configuring client rack awareness, producers send
requests to brokers in their local AZ, eliminating cross-AZ network traffic for
produce requests.

### Configuration

1. Configure preferred zones to match the AZs where your producers run.

1. Configure each producer's `client.id` to include the `diskless_az` pattern:

   ```properties
   # Producer in use1-az1
   client.id=my-producer,diskless_az=use1-az1

   # Producer in use1-az2
   client.id=my-producer,diskless_az=use1-az2
   ```

   The `diskless_az` value must match the broker's `broker.rack` configuration,
   which corresponds to the zone IDs in your `preferred_zones` setting.

1. Apply the same pattern to consumers for consistent local routing:

   ```properties
   # Consumer in use1-az1
   client.id=my-consumer,diskless_az=use1-az1
   ```

### Benefits

- **Eliminate most cross-AZ data transfer costs** for diskless topics
- **Improve cache hit rates** by routing clients to consistent brokers per partition
- **Works with any Kafka client version** since it uses the standard `client.id`
  property

:::note

> This `client.id` pattern is specific to Inkless Kafka with diskless topics. For

Reviewer: ```suggestion
This `client.id` pattern is specific to Kafka with diskless topics. For
```

  }
}
```

</TabItem>
</Tabs>

## How preferred zones work

When you configure preferred zones:

1. **New nodes** are placed in the specified zones when capacity is available.
1. **Zone validation** rejects invalid zone IDs at configuration time.
1. **Fallback behavior**: If a preferred zone is temporarily unavailable (for example,
   due to capacity constraints), nodes are placed in other available zones to maintain
1. **Existing nodes** are not automatically rebalanced, unless they meet the conditions
described below. Preferred zone creation is re-attempted on the next node recreation,
 such as during maintenance or plan changes.

### Minimum zone count

For high availability, configure at least three preferred zones. Configuring fewer than
three zones will fail, and requires special account permissions to reduce fault tolerance.

### Interaction with single zone configuration

If both `preferred_zones` and [`single_zone.availability_zone`](/docs/products/kafka/reference/advanced-params-inkless#single_zone.availability_zone) are configured,
the `single_zone` setting takes precedence when `single_zone.enabled` is true.

## Automatic rebalancing for Standard plans

> For Kafka services with Standard professional plans, nodes placed outside preferred

Reviewer 2: How can users identify whether their service is on a **Standard professional plan** versus another professional plan? We currently do not use this distinction in the docs or UI, so the term might confuse users.

Author: The new plans will be released with the new network billing. So we should probably wait for this before officially releasing preferred_zones? wdyt @Stuzanna?

Reviewer: Id like to release it when its ready.
Perhaps we move some of these changes into the pricing updates draft instead

Reviewer: We can't mention Standard here. Probably not best to mention details here either as its not public information :) 

Jonas is it exclusively the kafka-professional plans that do the auto-rebalance? No way for it to target BYOC or other diskless enabled plans?

Author: I added the filter that you gave me :-) https://github.com/aiven/aiven-core/pull/75192#discussion_r3468292742

Author: IMO auto rebalancing should be covered in docs though since otherwise users will wonder why it behaves like it does. What can we mention here?

Author: FYI preferred_zones was finally merged today so will be deployed soon

Reviewer: Yes I mean Standard plans is not public info.

So this applies to Inkless plans can be the way to describe it for now @harshini-rangaswamy and we can update to Standard later.

Reviewer 2: Updated the content to avoid mentioning Standard plans for now. I kept the rebalancing behavior high level. Could you take another look?

Reviewer: I mentioned plan names explicitly

For high availability, configure at least three preferred zones.

Configuring fewer than three zones reduces fault tolerance and requires special
account permissions.

### Interaction with single-zone configuration

If you configure both `preferred_zones` and
[`single_zone.availability_zone`](/docs/products/kafka/reference/advanced-params-inkless#single_zone.availability_zone)
settings and set `single_zone.enabled` to `true`, the `single_zone` setting takes
precedence.

## Automatic node rebalancing

When Aiven creates or replaces a node, it uses one of your preferred zones if
capacity is available. If none of the preferred zones have capacity, Aiven places the
node in another availability zone in the same region to keep your service available.

Aiven sends an alert when a node is running outside its preferred zone.

For Kafka plans that support automatic rebalancing, Aiven regularly checks for nodes
running outside their preferred zones. When capacity is available, Aiven automatically
moves those nodes back to a preferred zone.

For Kafka plans that don't support automatic rebalancing, move the node back to a
preferred zone manually from the Aiven Console or with the Aiven API.

> 

Reviewer: ```suggestion
Supported plans are:
* inkless-professional plans
* business and premium, inkless plans on BYOC
* kafka-professional plans
```

Author: Note: there is not really a way for the user to move the node manually. They just have to wait for the next maintenance or somehow force a genpump with plan change or turn off and on again.

### Minimum zone count

For high availability, configure at least three preferred zones.

Configuring fewer than three zones reduces fault tolerance and requires special
account permissions.

### Interaction with single-zone configuration

If you configure both `preferred_zones` and
[`single_zone.availability_zone`](/docs/products/kafka/reference/advanced-params-inkless#single_zone.availability_zone)
settings and set `single_zone.enabled` to `true`, the `single_zone` setting takes
precedence.

## Automatic node rebalancing

When Aiven creates or replaces a node, it uses one of your preferred zones if
capacity is available. If none of the preferred zones have capacity, Aiven places the
node in another availability zone in the same region to keep your service available.

Aiven sends an alert when a node is running outside its preferred zone.

For Kafka plans that support automatic rebalancing, Aiven regularly checks for nodes
running outside their preferred zones. When capacity is available, Aiven automatically
moves those nodes back to a preferred zone.

For Kafka plans that don't support automatic rebalancing, move the node back to a

> preferred zone manually from the Aiven Console or with the Aiven API.

Author: Nodes can't really be moved manually. Only way to do is to force a genbump. So for plans that don't support auto rebalance you'd just have to wait for the next maintenance etc.  cc @harshini-rangaswamy

Reviewer 2: Thanks! Updated this to remove the manual move wording.

- The service validates zone IDs when you save the configuration.
- If a preferred zone is unavailable, Aiven can use another zone in the same region
  to keep the service available. See
  [Automatic node rebalancing](#automatic-node-rebalancing).
- Existing nodes do not move immediately. Preferred zones take effect when Aiven
  recreates nodes, such as during maintenance or a plan change.

### Minimum zone count

For high availability, configure at least three preferred zones.

Configuring fewer than three zones reduces fault tolerance and requires special
account permissions.

### Interaction with single-zone configuration

If you configure both `preferred_zones` and
[`single_zone.availability_zone`](/docs/products/kafka/reference/advanced-params-inkless#single_zone.availability_zone)
settings and set `single_zone.enabled` to `true`, the `single_zone` setting takes
precedence.

## Automatic node rebalancing

When Aiven creates or replaces a node, it uses one of your preferred zones if
capacity is available. If none of the preferred zones have capacity, Aiven places the
node in another availability zone in the same region to keep your service available.

> Aiven sends an alert when a node is running outside its preferred zone.

Author: Think i would remove this. We don't send alerts for everything just if we think this is something actionable like plan not supported in AZ.
